packages/tooltip/src/Tooltip.tsx

```diff
@@ -0,0 +1,291 @@
+>
+  I am an uncontrolled Tooltip!
+</Tooltip>
+```
+ * @param props.children Content to appear inside of Tooltip.
+ * @param props.open Boolean to describe whether or not Tooltip is open.
+ * @param props.setOpen Callback to change the open state of the Tooltip.
+ * @param props.variant Whether the Tooltip should be `dark` or `light`.
+ * @param props.className Classname applied to Tooltip.
+ * @param props.align Alignment of Tooltip relative to trigger: `top`, `bottom`, `left`, `right`.
+ * @param props.justify Justification of Tooltip relative to trigger: `start`, `middle`, `end`.
+ * @param props.trigger Trigger element can be ReactNode or function.
+ * @param props.triggerEvent Whether the Tooltip should be triggered by a `click` or `hover`.
+ * @param props.id id given to Tooltip content.
+ */
+function Tooltip({
+  open: controlledOpen,
+  setOpen: controlledSetOpen,
+  className,
+  children,
+  trigger,
+  variant = Variant.Light,
+  triggerEvent = TriggerEvent.Hover,
+  align = 'top',
+  justify = 'start',
+  id,
+  shouldClose,
+  ...rest
+}: TooltipProps) {
+  const isControlled = !!controlledSetOpen;
+  const [uncontrolledOpen, uncontrolledSetOpen] = useState(false);
+  const open = isControlled ? controlledOpen : uncontrolledOpen;
+  // typescript is not recognizing isControlled checks that controlledSetOpen exists
+  const setOpen = isControlled
+    ? (controlledSetOpen as Function)
+    : uncontrolledSetOpen;
+
+  const triggerRef = useRef<HTMLElement>(null);
+  const tooltipRef = useRef<HTMLDivElement>(null);
+
+  const tooltipId = useMemo(
+    () => id || `tooltip-${Math.floor(Math.random() * Math.floor(10))}`,
+    [id],
+  );
+
+  const mapToTriggerEvent = (triggerType: TriggerEvent, triggerProps?: any) => {
+    if (triggerType === TriggerEvent.Hover) {
+      return {
+        onMouseEnter: debounce(() => {
+          setOpen(!open);
+        }, 35),
+        onMouseLeave: debounce(handleClose, 35),
+        onFocus: () => setOpen(true),
+        onBlur: handleClose,
+      };
+    }
+
+    if (triggerProps.onClick) {
+      return {
+        onClick: () => {
+          triggerProps.onClick();
+          setOpen(!open);
+        },
+      };
+    }
+
+    return {
+      onClick: () => setOpen(!open),
+    };
+  };
+
+  const handleClose = () => {
+    if (!shouldClose || shouldClose()) {
+      setOpen(false);
+    }
+  };
+
+  useHandleEscape(handleClose);
+
+  const handleBackdropClick = (e: MouseEvent) => {
+    const tooltipReference = tooltipRef && tooltipRef.current;
+
+    if (
+      triggerEvent === 'click' &&
+      tooltipReference &&
+      !tooltipReference.contains(e.target as HTMLElement)
+    ) {
+      handleClose();
+    }
+  };
+
+  useEventListener('click', handleBackdropClick, { enabled: open });
+
+  const triggerRect =
+    triggerRef &&
+    triggerRef.current &&
+    triggerRef.current.getBoundingClientRect();
+
+  // We are doing this to get the final alignment and justification from Popover
```

Comment:
I'm wondering if the better way to do this is by having Popover accept a function as a child and we pass in final values into the function.

Thoughts? cc @DesignerDave

Reply:
Huh. I really like that idea actually, and I don't think it would take all that much time! We're definitely gonna run up against this again...

Did you envision the function's signature being `(alignment, justification) => {}` or `({alignment, justification}) => {}`? The latter seems less clean, but more extensible.

Reply:
Generally I prefer the latter. Extensible and self documenting.

Reply:
Fair enough! @brookescarlett How does this sound?

```js
<Popover>
  {({alignment, justification}) => (
    <>
      <div>My alignment is {alignment}</div>
      <div>My justification is {justification}</div>
    </>
  )}
</Popover>
```

Reply:
It works 🔥This is sick